Improve Exception Handling

npwalker opened the pull request:
These changes move exception handling closer to the errors we care about and setup an error and error_count key to be included in the output for use in visualization tools.

npwalker committed:
Add exception handling when failing to retrieve a puppetdb metric

Prior to this commit, if a single metric fails to be retrieved
from the api then the whole metric script would fail.

This commit adds exception handling that causes the error received
to be logged in place of the metric hash.

Fixes #9

npwalker committed:
Add exception handling to puppet server mertrics script

Prior to this commit, if the puppetserver status endpoint wasn't
avialable or didn't respond then no metrics file would be created.

After this commit, a metrics file is created with the error message
in place of the metrics hash.

Fixes #9

npwalker committed:
Create method for retrieving the status endpoint

This allows for raising the error at the correct place, when
gathering the metrics, and then still get the api start and stop
times recorded.

npwalker commented:
Example PuppetDB output:

~~~
[root@master201710-centos pe_metric_curl_cron_jobs]# cat /opt/puppetlabs/pe_metric_curl_cron_jobs/puppetdb/127.0.0.1/20170317T164532Z.json 
{
  "timestamp": "2017-03-17T16:45:32Z",
  "servers": {
    "127-0-0-1": {
      "puppetdb": {
        "api-query-start": "2017-03-17T16:45:32Z",
        "commands_queue": {
        },
        "catalog_hash_miss": {
          "OneMinuteRate": 0.0,
          "MeanRate": 0.0,
          "FifteenMinuteRate": 0.0,
          "Max": 0.0,
          "50thPercentile": 0.0,
          "Mean": 0.0,
          "DurationUnit": "milliseconds",
          "95thPercentile": 0.0,
          "99thPercentile": 0.0,
          "98thPercentile": 0.0,
          "Min": 0.0,
          "999thPercentile": 0.0,
          "RateUnit": "events/second",
          "75thPercentile": 0.0,
          "Count": 0,
          "StdDev": 0.0,
          "FiveMinuteRate": 0.0
        },
        "catalog_hash_match": {
          "OneMinuteRate": 1.2934867175847721e-36,
          "MeanRate": 0.0005300756235879249,
          "FifteenMinuteRate": 6.837481262924639e-06,
          "Max": 14.53112,
          "50thPercentile": 1.5930279999999999,
          "Mean": 1.593027999999671,
          "DurationUnit": "milliseconds",
          "95thPercentile": 1.5930279999999999,
          "99thPercentile": 1.5930279999999999,
          "98thPercentile": 1.5930279999999999,
          "Min": 1.2178609999999999,
          "999thPercentile": 1.5930279999999999,
          "RateUnit": "events/second",
          "75thPercentile": 1.5930279999999999,
          "Count": 35,
          "StdDev": 2.4159365060802935e-07,
          "FiveMinuteRate": 5.034068607111962e-10
        },
        "replace_catalog_time": {
          "OneMinuteRate": 1.2934867175847721e-36,
          "MeanRate": 0.0005300745804729688,
          "FifteenMinuteRate": 6.837481262924639e-06,
          "Max": 196.142147,
          "50thPercentile": 152.41414699999999,
          "Mean": 152.41414700000914,
          "DurationUnit": "milliseconds",
          "95thPercentile": 152.41414699999999,
          "99thPercentile": 152.41414699999999,
          "98thPercentile": 152.41414699999999,
          "Min": 88.754767,
          "999thPercentile": 152.41414699999999,
          "RateUnit": "events/second",
          "75thPercentile": 152.41414699999999,
          "Count": 35,
          "StdDev": 6.7409838702997155e-06,
          "FiveMinuteRate": 5.034068607111962e-10
        },
        "replace_facts_time": {
          "OneMinuteRate": 1.0949128678242564e-36,
          "MeanRate": 0.0005300735577156564,
          "FifteenMinuteRate": 6.7619297567682855e-06,
          "Max": 107.223889,
          "50thPercentile": 70.48064699999999,
          "Mean": 70.48064700003258,
          "DurationUnit": "milliseconds",
          "95thPercentile": 70.48064699999999,
          "99thPercentile": 70.48064699999999,
          "98thPercentile": 70.48064699999999,
          "Min": 58.457919,
          "999thPercentile": 70.48064699999999,
          "RateUnit": "events/second",
          "75thPercentile": 70.48064699999999,
          "Count": 35,
          "StdDev": 2.376520286155274e-05,
          "FiveMinuteRate": 4.869032207729713e-10
        },
        "store_report_time": {
          "OneMinuteRate": 1.6608698215268455e-36,
          "MeanRate": 0.0005300724291264567,
          "FifteenMinuteRate": 6.952395958267435e-06,
          "Max": 147.675312,
          "50thPercentile": 87.224291,
          "Mean": 87.22429099994196,
          "DurationUnit": "milliseconds",
          "95thPercentile": 87.224291,
          "99thPercentile": 87.224291,
          "98thPercentile": 87.224291,
          "Min": 55.888076,
          "999thPercentile": 87.224291,
          "RateUnit": "events/second",
          "75thPercentile": 87.224291,
          "Count": 35,
          "StdDev": 4.2639668925693235e-05,
          "FiveMinuteRate": 5.292170823830724e-10
        },
        "global_retried": {
          "RateUnit": "events/second",
          "Count": 0,
          "FifteenMinuteRate": 0.0,
          "FiveMinuteRate": 0.0,
          "MeanRate": 0.0,
          "OneMinuteRate": 0.0
        },
        "global_retry_counts": {
          "Max": 0,
          "50thPercentile": 0.0,
          "Mean": 0.0,
          "95thPercentile": 0.0,
          "99thPercentile": 0.0,
          "98thPercentile": 0.0,
          "Min": 0,
          "999thPercentile": 0.0,
          "75thPercentile": 0.0,
          "Count": 105,
          "StdDev": 0.0
        },
        "global_retry_persistence_time": {
        },
        "replace_catalog_retried": {
          "RateUnit": "events/second",
          "Count": 0,
          "FifteenMinuteRate": 0.0,
          "FiveMinuteRate": 0.0,
          "MeanRate": 0.0,
          "OneMinuteRate": 0.0
        },
        "replace_catalog_retry_counts": {
          "Max": 0,
          "50thPercentile": 0.0,
          "Mean": 0.0,
          "95thPercentile": 0.0,
          "99thPercentile": 0.0,
          "98thPercentile": 0.0,
          "Min": 0,
          "999thPercentile": 0.0,
          "75thPercentile": 0.0,
          "Count": 35,
          "StdDev": 0.0
        },
        "replace_catalog_retry_persistence_time": {
        },
        "replace_facts_retried": {
          "RateUnit": "events/second",
          "Count": 0,
          "FifteenMinuteRate": 0.0,
          "FiveMinuteRate": 0.0,
          "MeanRate": 0.0,
          "OneMinuteRate": 0.0
        },
        "replace_facts_retry_counts": {
          "Max": 0,
          "50thPercentile": 0.0,
          "Mean": 0.0,
          "95thPercentile": 0.0,
          "99thPercentile": 0.0,
          "98thPercentile": 0.0,
          "Min": 0,
          "999thPercentile": 0.0,
          "75thPercentile": 0.0,
          "Count": 35,
          "StdDev": 0.0
        },
        "replace_facts_retry_persistence_time": {
        },
        "store_report_retried": {
          "RateUnit": "events/second",
          "Count": 0,
          "FifteenMinuteRate": 0.0,
          "FiveMinuteRate": 0.0,
          "MeanRate": 0.0,
          "OneMinuteRate": 0.0
        },
        "store_reports_retry_counts": {
          "Max": 0,
          "50thPercentile": 0.0,
          "Mean": 0.0,
          "95thPercentile": 0.0,
          "99thPercentile": 0.0,
          "98thPercentile": 0.0,
          "Min": 0,
          "999thPercentile": 0.0,
          "75thPercentile": 0.0,
          "Count": 35,
          "StdDev": 0.0
        },
        "store_report_retry_persistence_time": {
        },
        "error": [
          "757: unexpected token at '\"No mbean 'org.apache.activemq:type=Broker,brokerName=localhost,destinationType=Queue,destinationName=puppetlabs.puppetdb.commands' found\"'",
          "757: unexpected token at '\"No mbean 'puppetlabs.puppetdb.mq:name=global.retry-persistence-time' found\"'",
          "757: unexpected token at '\"No mbean 'puppetlabs.puppetdb.mq:name=replace catalog.9.retry-persistence-time' found\"'",
          "757: unexpected token at '\"No mbean 'puppetlabs.puppetdb.mq:name=replace facts.5.retry-persistence-time' found\"'",
          "757: unexpected token at '\"No mbean 'puppetlabs.puppetdb.mq:name=store report.8.retry-persistence-time' found\"'"
        ],
        "error_count": 5,
        "api-query-duration": 2.150136731
      }
    }
  }
}
~~~

Example Puppet Server output:

~~~
[root@master201710-centos pe_metric_curl_cron_jobs]# cat /opt/puppetlabs/pe_metric_curl_cron_jobs/puppetserver/127.0.0.1/20170317T163101Z.json 
{
  "timestamp": "2017-03-17T16:31:01Z",
  "servers": {
    "127-0-0-1": {
      "puppetserver": {
        "error": "Connection refused - connect(2) for \"127.0.0.1\" port 8140",
        "error_count": 1,
        "api-query-start": "2017-03-17T16:31:01Z",
        "api-query-duration": 0.025065816
      }
    }
  }
}
~~~

~~~
[root@master201710-centos pe_metric_curl_cron_jobs]# cat /opt/puppetlabs/pe_metric_curl_cron_jobs/puppetserver/127.0.0.1/20170317T165001Z.json 
{
  "timestamp": "2017-03-17T16:50:01Z",
  "servers": {
    "127-0-0-1": {
      "puppetserver": {
        "pe-puppet-profiler": {
          "service_version": "2017.1.0.5",
          "service_status_version": 1,
          "detail_level": "debug",
          "state": "running",
          "status": {
            "experimental": {
              "function-metrics": [

              ],
              "resource-metrics": [

              ],
              "catalog-metrics": [

              ],
              "inline-metrics": [

              ]
            }
          },
          "active_alerts": [

          ]
        },
        "pe-jruby-metrics": {
          "service_version": "2017.1.0.5",
          "service_status_version": 1,
          "detail_level": "debug",
          "state": "running",
          "status": {
            "experimental": {
              "jruby-pool-lock-status": {
                "current-state": ":not-in-use",
                "last-change-time": "2017-03-17T16:32:09.475Z"
              },
              "metrics": {
                "average-lock-wait-time": 0,
                "num-free-jrubies": 1,
                "borrow-count": 4,
                "average-requested-jrubies": 0.0,
                "borrow-timeout-count": 0,
                "return-count": 4,
                "borrow-retry-count": 0,
                "borrowed-instances": [

                ],
                "average-borrow-time": 1662,
                "num-jrubies": 1,
                "requested-count": 4,
                "average-lock-held-time": 0,
                "requested-instances": [

                ],
                "average-free-jrubies": 0.9999999565881191,
                "num-pool-locks": 0,
                "average-wait-time": 1
              }
            }
          },
          "active_alerts": [

          ]
        },
        "pe-master": {
          "service_version": "2017.1.0.5",
          "service_status_version": 1,
          "detail_level": "debug",
          "state": "running",
          "status": {
            "experimental": {
              "http-metrics": [
                {
                  "route-id": "total",
                  "count": 4,
                  "mean": 1669,
                  "aggregate": 6676
                },
                {
                  "route-id": "puppet-v3-environment_classes-/*/",
                  "count": 2,
                  "mean": 3218,
                  "aggregate": 6436
                },
                {
                  "route-id": "puppet-v3-environments",
                  "count": 2,
                  "mean": 25,
                  "aggregate": 50
                },
                {
                  "route-id": "puppet-v3-static_file_content-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-file_content-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-file_metadata-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-experimental",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "other",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-status-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-experimental-metrics-dashboard-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-report-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-experimental-dashboard_html",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-resource_types-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-node-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-catalog-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-resource_type-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-file_metadatas-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-file_bucket_file-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-experimental-js-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                },
                {
                  "route-id": "puppet-v3-environment-/*/",
                  "count": 0,
                  "mean": 0,
                  "aggregate": 0
                }
              ]
            }
          },
          "active_alerts": [

          ]
        },
        "status-service": {
          "service_version": "0.7.1",
          "service_status_version": 1,
          "detail_level": "debug",
          "state": "running",
          "status": {
            "experimental": {
              "jvm-metrics": {
                "heap-memory": {
                  "committed": 175112192,
                  "init": 134217728,
                  "max": 201326592,
                  "used": 114758848
                },
                "non-heap-memory": {
                  "committed": 281411584,
                  "init": 2555904,
                  "max": -1,
                  "used": 174470528
                },
                "file-descriptors": {
                  "used": 34,
                  "max": 4096
                },
                "gc-stats": {
                  "G1 Young Generation": {
                    "count": 108,
                    "total-time-ms": 1619
                  },
                  "G1 Old Generation": {
                    "count": 0,
                    "total-time-ms": 0
                  }
                },
                "cpu-usage": 0.04999000206589699,
                "gc-cpu-usage": 0.0,
                "up-time-ms": 1087587,
                "start-time-ms": 1489768313792
              }
            }
          },
          "active_alerts": [

          ]
        },
        "error": "",
        "error_count": 0,
        "api-query-start": "2017-03-17T16:50:01Z",
        "api-query-duration": 0.139964373
      }
    }
  }
}
~~~

reidmv commented on templates/puppetserver_metrics.epp:
```diff
@@ -18,14 +18,23 @@ HOSTS = [
+  metrics = open(uri, {ssl_verify_mode: OpenSSL::SSL::VERIFY_NONE}).read
+  data = JSON.parse(metrics)
+  data = data.merge({"error" => "", "error_count" => 0})
+  rescue Exception => e
```
Looks like the `begin; ...; rescue; ...; end` structure got broken here when the method was split out. Needs to be fixed

npwalker replied:
I was pretty much just following this: http://seejohncode.com/2012/04/17/short-tip-rescuing-a-method-in-ruby/

reidmv replied:
Ah, cool. Then this is just an indentation issue here; `rescue` should be at the same indentation level as `def`.

reidmv replied:
Besides fixing the indentation here, the `begin; ...; end` back under the `HOSTS.each` block should be removed if we aren't catching exceptions there.

reidmv reviewed (changes requested):
Begin/Rescue is broken; needs fixed. Begin and Rescue need to be together, not split across methods.

npwalker committed:
Collect all PuppetDB metric errors into an array for output

npwalker committed:
Add error and error count keys

In PuppetServer metrics you can only fail to collect the status
endpoint so there's one error and error count key.

In PuppetDB metrics you can fail to gather any individual metric
so there is an error and error count key per metric.

npwalker commented:
@reidmv all right anything else?
